src/tsconfig/validator.ts

```diff
@@ -0,0 +1,407 @@
+ * This can be used to generate synthetics values that would match the matcher.
+ * It is not guaranteed that hints are received or that hints are complete.
+ *
+ * @returns true if the value matches
+ */
+type Matcher = (value: any, reporter: (message: string) => void, hints?: (allowed: any[]) => void) => boolean;
+
+export enum RuleType {
+  PASS,
+  FAIL,
+}
+
+export interface RuleSetOptions {
+  /**
+   * Defines the behavior for any encountered fields for which no rules are defined.
+   * The default is to pass these fields without validation,
+   * but this can also be set to fail any unexpected fields.
+   *
+   * @default RuleType.PASS
+   */
+  readonly unexpectedFields: RuleType;
+}
+
+interface Rule {
+  field: string;
+  type: RuleType;
+  matcher: Matcher;
+}
+
+export class RuleSet {
+  private _rules: Array<Rule> = [];
+  public get rules(): Array<Rule> {
+    return this._rules;
+  }
+
+  /**
+   * Return all fields for which a rule exists
+   */
+  public get fields(): Array<string> {
+    return [...new Set(this._rules.map((r) => r.field))];
+  }
+
+  /**
+   * Return a list of fields that are allowed, or undefined if all are allowed.
+   */
+  public get allowedFields(): Array<string> | undefined {
+    if (this.options.unexpectedFields === RuleType.FAIL) {
+      return this.fields;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Find all required fields by evaluating every rule in th set against undefined.
+   * If the rule fails, the key must be required.
+   *
+   * @returns A list of keys that must be included or undefined
+   */
+  public get requiredFields(): Array<string> {
+    const required: string[] = [];
+    const noop = () => {};
+
+    for (const rule of this._rules) {
+      const key = rule.field;
+      const matcherResult = rule.matcher(undefined, noop);
+
+      switch (rule.type) {
+        case RuleType.PASS:
+          if (!matcherResult) {
+            required.push(key);
+          }
+          break;
+        case RuleType.FAIL:
+          if (matcherResult) {
+            required.push(key);
+          }
+          break;
+        default:
+          continue;
+      }
+    }
+
+    return required;
+  }
+
+  public constructor(
+    public readonly options: RuleSetOptions = {
+      unexpectedFields: RuleType.PASS,
+    },
+  ) {}
+
+  /**
+   * Requires the matcher to pass for the given field.
+   * Otherwise a violation is detected.
+   *
+   * @param field The field the rule applies to
+   * @param matcher The matcher function
+   */
+  public pass(field: string, matcher: Matcher) {
+    this._rules.push({ field, matcher, type: RuleType.PASS });
+  }
+
+  /**
+   * Requires the matcher to pass for the given field.
+   * Otherwise a violation is detected.
+   *
+   * @param data The data object we want to add rules for.
+   * @param caseSensitive Whether the string matcher should be case sensitive or not.
+   */
+  public passMany(data: object, caseSensitive = false) {
```

Comment:
Can this `object` not be typed a bit more explicitly? 😇

Reply:
`caseSensitive` is a bit weird to me here. Can't we just leave it out, have case insensitive matching as default, and allow passing a `Matcher` as well as one of the values? So that if users want a case insensitive match, they can pass `Match.strEq('asdf', true)`, or even just `Match.strictStrEq('asdf')` ?

Reply:
I got rid of passMany/failMany. It was a bit weird that it was responsible for the conversion from value to Matcher. We might want to re-introduce this later as something that takes `Record<string, Matcher>` to make it easier to define many rules. But for now this will do.